nomos-services/membership/src/backends/mock.rs

```diff
@@ -0,0 +1,439 @@
+        block_number: BlockNumber,
+    ) -> Result<MembershipProviders, MembershipBackendError> {
+        let k = self
+            .settings
+            .settings_per_service
+            .get(&service_type)
+            .ok_or_else(|| MembershipBackendError::Other("Service type not found".into()))?;
+        let index = block_number.saturating_sub(k.historical_block_delta);
+        return Ok(self.get_snapshot(index, service_type));
+    }
+
+    async fn get_latest_providers(
+        &self,
+        service_type: ServiceType,
+    ) -> Result<MembershipProviders, MembershipBackendError> {
+        return Ok(self.get_snapshot(self.latest_block_number, service_type));
+    }
+
+    async fn update(
+        &mut self,
+        update: FinalizedBlockEvent,
+    ) -> Result<HashMap<ServiceType, MembershipProviders>, MembershipBackendError> {
+        let block_number = update.block_number;
+        let mut result = HashMap::new();
+
+        let mut current_entry = self
+            .membership
+            .get(&self.latest_block_number)
+            .cloned()
+            .unwrap_or_default();
+
+        for (provider_info, declaration_update) in update.updates {
+            if !self
+                .settings
+                .settings_per_service
+                .contains_key(&declaration_update.service_type)
+            {
+                continue;
+            }
+
+            let service_type = declaration_update.service_type;
+            let service_data = current_entry.entry(service_type).or_default();
+            service_data.insert(provider_info, declaration_update.clone());
+            result.insert(service_type, HashMap::new());
+        }
+
+        self.latest_block_number = block_number;
+        // todo: figure out pruning the old entries
+        self.membership.insert(block_number, current_entry.clone());
+
+        for (service_type, snapshot) in current_entry {
+            result.insert(
+                service_type,
+                snapshot
+                    .iter()
+                    .map(|(provider_info, declaration_update)| {
+                        (
+                            provider_info.provider_id,
+                            declaration_update.locators.clone(),
+                        )
+                    })
+                    .collect(),
+            );
+        }
+
+        Ok(result)
+    }
+}
+
+impl MockMembershipBackend {
+    fn get_snapshot(
+        &self,
+        block_number: BlockNumber,
+        service_type: ServiceType,
+    ) -> MembershipProviders {
+        self.membership
+            .get(&block_number)
+            .and_then(|entry| entry.get(&service_type))
+            .map_or_else(HashMap::new, |data| {
+                data.iter()
+                    .map(|(provider_info, declaration_update)| {
+                        (
+                            provider_info.provider_id,
+                            declaration_update.locators.clone(),
```

Comment:
For mock this will work, when implementing with storage, we just need to keep provider_id, it's active block numbers and latest locators.

Reply:
For membership service I don't think we need to know service parameters - sdp will determine the state of provider and return the ProviderUpdate. SDP will manage service parameters (updating these parameters is not yet specified as far as I know). Pruning for very old data here should be done by block number, configured separately via membership settings.